### PR TITLE
Expand cache scorecard proto and add dedicated RPC for cache scorecard

### PR DESCRIPTION
**Pull request by bduffany**
Expanding the cache scorecard result message with more details, and adding a new field `result` which will contain a flat results list. This flattened list is easiest for the client to slice and dice.

Might need to iterate on this in subsequent PRs if we see that this approach doesn't scale well, but this flat list works well for development purposes. The plan is to only populate the new fields behind a flag.

---

**Version bump**: Patch

### CODE REVIEW

File: proto/cache.proto [hunk 1]
```diff
@@ -30,13 +36,85 @@ message CacheStats {
...
 }
 
+// Request to retrieve detailed per-request cache stats.
+message GetCacheScoreCardRequest {
+  context.RequestContext request_context = 1;
+
+  // The invocation ID for which to look up cache stats.
+  string invocation_id = 2;
+
+  // TODO(bduffany): Add paging
```
Comment (tylerwilliams) [inline proto/cache.proto:46]: nit: i think we have page token protos -- can you just copy those from elsewhere? should be the same.

File: proto/cache.proto [hunk 2]
```diff
@@ -30,13 +36,85 @@ message CacheStats {
...
+    // Whether the transferred payload was compressed.
+    bool compressed = 10;
+
+    // The compressed size of the transferred payload. Should be ignored if
+    // `compressed` is false.
+    int64 compressed_size_bytes = 11;
   }
 
   // In the interest of saving space, we only show cache misses.
+  // TODO(bduffany): use flat `results` list and deprecate this
```
Comment (tylerwilliams) [inline proto/cache.proto:116]: Something to consider: in the interest of performance we might want to do something in the future where by default we just show misses, and then if you hit a button in the UI, we go re-fetch the full scorecard.

Just putting this here as a note so we can both think about what that might look like -- if this list contained only miss results.

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: proto/cache.proto (hunk 1 above)
Comment (bduffany) [reply proto/cache.proto]: The only page token proto I could find was in `target_pagination.proto` and the fields will be different. Added a basic page token here

File: proto/cache.proto (hunk 2 above)
Comment (bduffany) [reply proto/cache.proto]: Ack, sg to return only misses by default. Will think about that when adding the sorting/filtering fields

File: proto/cache.proto [hunk 3]
```diff
@@ -30,11 +36,96 @@ message CacheStats {
...
+
+message GetCacheScoreCardResponse {
+  context.ResponseContext response_context = 1;
+
+  // The cache results for the current page.
+  repeated ScoreCard.Result results = 2;
+
+  // A token used to fetch more results. If empty, there are no results
+  // remaining.
+  string page_token = 3;
```
Comment (tylerwilliams) [inline proto/cache.proto:69]: nit: call this next_page_token (to allow us to add prev_page_token later if we want)

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: proto/cache.proto (hunk 3 above)
Comment (bduffany) [reply proto/cache.proto:69]: Done

File: proto/cache.proto [hunk 4]
```diff
@@ -30,11 +36,96 @@ message CacheStats {
...
+  // The invocation ID for which to look up cache stats.
+  string invocation_id = 2;
+
+  // Pagination token. Not intended for direct client usage.
+  message PageToken {
+    // The start index in the scorecard results list, after sorting and
+    // filtering results. Number of results to fetch is determined by the
+    // server.
+    int64 index = 1;
+  }
```
Comment (tylerwilliams) [inline proto/cache.proto:52]: Can we put this in context.proto (or make pagination.proto)?

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: proto/cache.proto (hunk 4 above)
Comment (bduffany) [reply proto/cache.proto]: Done

**Review by tylerwilliams — APPROVED** [review]

**Pull request merged**